Library/Homebrew/formula_installer.rb

```diff
@@ -393,7 +393,8 @@ def expand_dependencies(deps)
 
       if (dep.optional? || dep.recommended?) && build.without?(dep)
         Dependency.prune
-      elsif dep.build? && install_bottle_for?(dependent, build)
+      elsif dep.build? && (install_bottle_for?(dependent, build) ||
+           !ARGV.build_formula_from_source?(dependent))
```

Comment:
May be worth pulling this and other `(...||...)` into a separate function or variable.